### PR TITLE
Fix missed clustermesh config change race condition with back-to-back changes

### LINKED ISSUE #24491: CI: IntegrationTests: ClusterMeshTestSuite.TestWatchConfigDirectory: timeout while waiting for changed event
Observed in https://github.com/cilium/cilium/actions/runs/4471325368/jobs/7856059453?pr=24477

```
----------------------------------------------------------------------
FAIL: config_test.go:69: ClusterMeshTestSuite.TestWatchConfigDirectory

config_test.go:178:
    expectChange(c, cm, "cluster2")
config_test.go:46:
    c.Fatal("timeout while waiting for changed event")
... Error: timeout while waiting for changed event
```

[logs_1265704.zip](https://github.com/cilium/cilium/files/11026575/logs_1265704.zip)

### PR DESCRIPTION
**Pull request by giorio94**
Currently, the config watcher used to detect changes in the clustermesh configuration is affected by a possible race condition in case the same file is modified multiple times back-to-back, causing one of the updates to be missed. Specifically, there is a small chance that the file content changes between when it is read and when the watcher is established, hence losing that notification. This PR fixes this by re-reading the file after having established the watcher, to ensure that we are always processing the most up-to-date version of the file, and that subsequent modifications are guaranteed to be detected.

This issue could realistically occur only in tests, due the tight timing constraints (hence I'm adding the `misc` release note rather than `bug`), and it caused a seldom failure of the TestWatchConfigDirectory test with the following error:

FAIL: config_test.go:69: ClusterMeshTestSuite.TestWatchConfigDirectory

config_test.go:178:
    expectChange(c, cm, "cluster2")
config_test.go:46:
    c.Fatal("timeout while waiting for changed event")
... Error: timeout while waiting for changed event

I've run the test locally with 1K+ repetitions and observed no failures with the fix.
Marking for backport in v1.13 to propagate the fix for the flaky test.

Fixes: #24491

```release-note
Fix missed clustermesh config change race condition with back-to-back changes
```

### CODE REVIEW

File: pkg/clustermesh/config.go [hunk 1]
```diff
@@ -98,9 +99,19 @@ func (cdw *configDirectoryWatcher) handle(abspath string) {
 		// when the underlying file gets updated, if path points to a symbolic link.
 		// This is required to correctly detect file modifications when the folder
 		// is mounted from a Kubernetes ConfigMap/Secret.
-		if err := cdw.watcher.Add(abspath); err != nil {
+		if err := cdw.watcher.Add(abspath); err != nil && !errors.Is(err, os.ErrNotExist) {
```
Comment (YutaroHayakawa) [inline pkg/clustermesh/config.go:102]: Nit: We can use `os.IsNotExist` for the same purpose.

**Review by YutaroHayakawa — APPROVED** [review]
LGTM with nit.

File: pkg/clustermesh/config.go (hunk 1 above)
Comment (giorio94) [reply pkg/clustermesh/config.go:102]: The godoc of `os.IsNotExist` mentions *"This function predates errors.Is. It only supports errors returned by the os package. New code should use errors.Is(err, fs.ErrNotExist)"*, hence the current approach should be preferred. `os.ErrNotExist` is an alias of `fs.ErrNotExist`.

**Pull request merged**